pkg/kubernetes/deployment/appsv1/v1alpha1/deployment.go

```diff
@@ -0,0 +1,257 @@
+//
+// This is the basic building block to create functional operations
+// against the deployment instance
+type deployBuildOption func(*deploy)
+
+// rolloutStatus  is a typed function that abstracts status message formation logic
+type rolloutStatus func(*deploy) string
+
+// deploy is a wrapper over aapi_apps_v1.Deployment
+type deploy struct {
+	object *api_apps_v1.Deployment // kubernetes deployment instance
+	checks []predicate             // predicate list for deploy
+}
+
+// predicateKey is wrapper over string. It is used to get predicate function and status msg.
+type predicateKey string
+
+// rolloutStatuses contains a group of status message for each predicate checks.
+// It useses predicateKey as key.
+var rolloutStatuses = map[predicateKey]rolloutStatus{
+	// ProgressDeadlineExceededPK refer to rolloutStatus for predicate IsProgressDeadlineExceeded.
+	ProgressDeadlineExceededPK: func(d *deploy) string {
+		return "Deployment exceeded its progress deadline"
+	},
+	// OlderReplicaActivePK refer to rolloutStatus for predicate IsOlderReplicaActive.
+	OlderReplicaActivePK: func(d *deploy) string {
+		if d.object.Spec.Replicas != nil {
+			return fmt.Sprintf("Waiting for deployment rollout to finish: %d out of %d new replicas have been updated",
+				d.object.Status.UpdatedReplicas, *d.object.Spec.Replicas)
+		}
+		return "Waiting for deployment rollout to finish: some older replicas have been updated"
+	},
+	// TerminationInProgressPK refer rolloutStatus for predicate IsTerminationInProgress.
+	TerminationInProgressPK: func(d *deploy) string {
+		return fmt.Sprintf("Waiting for deployment rollout to finish: %d old replicas are pending termination",
+			d.object.Status.Replicas-d.object.Status.UpdatedReplicas)
+	},
+	// UpdationInProgressPK refer to rolloutStatus for predicate IsUpdationInProgress.
+	UpdationInProgressPK: func(d *deploy) string {
+		return fmt.Sprintf("Waiting for deployment rollout to finish: %d of %d updated replicas are available",
+			d.object.Status.AvailableReplicas, d.object.Status.UpdatedReplicas)
+	},
+	// NotSyncSpecPK refer to status rolloutStatus for predicate IsNotSyncSpec.
+	NotSyncSpecPK: func(d *deploy) string {
+		return "Waiting for deployment spec update to be observed"
+	},
+}
+
+// rolloutChecks contains a group of predicate it useses predicateKey as key.
+var rolloutChecks = map[predicateKey]predicate{
+	// ProgressDeadlineExceededPK refer to predicate IsProgressDeadlineExceeded.
+	ProgressDeadlineExceededPK: IsProgressDeadlineExceeded(),
+	// OlderReplicaActivePK refer to predicate IsOlderReplicaActive.
+	OlderReplicaActivePK: IsOlderReplicaActive(),
+	// TerminationInProgressPK refer to predicate IsTerminationInProgress.
+	TerminationInProgressPK: IsTerminationInProgress(),
+	// UpdationInProgressPK refer to predicate IsUpdationInProgress.
+	UpdationInProgressPK: IsUpdationInProgress(),
+	// NotSyncSpecPK refer to predicate IsSyncSpec.
+	NotSyncSpecPK: IsNotSyncSpec(),
+}
+
+const (
```

Comment:
I was referring some external code base & this is the recommended naming
```go

type predicateName string

const (
        // PredicateNameProgressDeadlineExceeded refer to predicate IsProgressDeadlineExceeded.
	PredicateNameProgressDeadlineExceeded predicateName = "ProgressDeadlineExceeded"

	// PredicateNameSpecSynced refer to predicate IsSpecSynced
	PredicateNameSpecSynced predicateName = "SpecSynced"

	// PredicateNameOlderReplicaActive refer to predicate IsOlderReplicaActive
	PredicateNameOlderReplicaActive predicateName = "OlderReplicaActive"

	// PredicateNameTerminationInProgress refer to predicate IsTerminationInProgress
	PredicateNameTerminationInProgress predicateName = "TerminationInProgress"

	// PredicateNameUpdateInProgress refer to predicate IsUpdateInProgress.
	PredicateNameUpdateInProgress predicateName = "UpdateInProgress"
)
```

Reply:
done please check it once.